core/src/main/java/org/apache/calcite/sql/SqlBasicFunction.java

```diff
@@ -42,7 +42,7 @@
  * behavior only by providing strategy objects, not by overriding methods in a
  * subclass.
  */
-public class SqlBasicFunction extends SqlFunction {
+public final class SqlBasicFunction extends SqlFunction {
```

Comment:
this looks like a major change, which may break users of this class

Reply:
I'll remove it. I don’t remember why I added it—thanks for pointing it out.

Reply:
Ok, I remembered why I added it. If you look at the comment in the class:
```
The class is final, and instances are immutable

Since the class is final, you can modify behavior only by providing strategy objects, not by overriding methods in a subclass.
```
This comment was added when the class was introduced, and from the Git history, I can see that making it final was simply overlooked.

Think it is right to make it final.

Reply:
I agree that the intent was to have the class final, but if someone happens to have inherited from this class in some project that uses Calcite, they will now be stuck.

Reply:
@mihaibudiu This class can’t be inherited from outside because it has a private constructor, which prevents other classes from calling super(). Making it non-final only allows subclasses within SqlBasicFunction itself.

Perhaps this was done to allow adding an inner Builder class in the future, but I’m not sure.

Let’s ask the author of the commit, @julianhyde, about it

Reply:
Making constructor methods private seems to be a recent change. SqlBasicFunction was extended in the project I'm using, and this seems to be a thing to keep in mind when upgrading calcite.